scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -651,294 +667,149 @@ def dijkstra(csgraph, directed=True, indices=None,
-
-            # v has now been scanned: add the distance to the results
-            dist_matrix[i, v.index] = v.val
-
-    free(nodes)
+                                dist_matrix, pred, return_pred,
+                                sources, return_sources, limit)
     return 0
 
 @cython.boundscheck(False)
-cdef int _dijkstra_undirected_multi(
+cdef int _dijkstra_multi_separate(
             const int[:] source_indices,
             const double[:] csr_weights,
             const int[:] csr_indices,
             const int[:] csr_indptr,
             const double[:] csrT_weights,
             const int[:] csrT_indices,
             const int[:] csrT_indptr,
-            double[:] dist_matrix,
-            int[:] pred,
-            int[:] sources,
+            double[:, :] dist_matrix,
+            int[:, :] pred,
+            int[:, :] sources,
             DTYPE_t limit) except -1:
     cdef:
-        unsigned int N = dist_matrix.shape[0]
-        int return_pred = (pred.size > 0)
-        FibonacciHeap heap
-        FibonacciNode *v
-        FibonacciNode* nodes = <FibonacciNode*> malloc(N *
-                                                       sizeof(FibonacciNode))
-    if nodes == NULL:
-        raise MemoryError("Failed to allocate memory in "
-                          "_dijkstra_undirected_multi")
-
-    _dijkstra_setup_heap_multi(&heap, nodes, source_indices,
-                               sources, dist_matrix, return_pred)
-
-    while heap.min_node:
-        v = remove_min(&heap)
-        v.state = SCANNED
-
-        _dijkstra_scan_heap_multi(&heap, v, nodes,
-                                  csr_weights, csr_indices, csr_indptr,
-                                  pred, sources, return_pred, limit)
-
-        _dijkstra_scan_heap_multi(&heap, v, nodes,
-                                  csrT_weights, csrT_indices, csrT_indptr,
-                                  pred, sources, return_pred, limit)
+        unsigned int Nind = source_indices.shape[0]
+        unsigned int i
+        int source_list[1]
 
-        #v has now been scanned: add the distance to the results
-        dist_matrix[v.index] = v.val
+    if dist_matrix.shape[0] != Nind:
+        raise RuntimeError(
+            f"Not enough rows in distances matrix. Got {dist_matrix.shape[0]}, expected {Nind}."
+        )
+    if pred.shape[0] != Nind:
+        raise RuntimeError(
+            f"Not enough rows in predecessors matrix. Got {pred.shape[0]}, expected {Nind}."
+        )
+    if sources.shape[0] != Nind:
+        raise RuntimeError(
+            f"Not enough rows in sources matrix. Got {sources.shape[0]}, expected {Nind}."
+        )
 
-    free(nodes)
-    return 0
+    for i in range(Nind):
+        source_list[0] = source_indices[i]
+        _dijkstra(source_list,
+                  csr_weights, csr_indices, csr_indptr,
+                  csrT_weights, csrT_indices, csrT_indptr,
+                  dist_matrix[i], pred[i], sources[i], limit)
```

Comment:
Actually `_dijkstra` function does multiple sources (if `min_only=True`). However here we need to pass one source at a time to compute shortest distance from each of the index in `indices`. I tried implementing `_dijkstra_single_source` and remove the following loop, 

```python
for i in range(Nind):
    j_source = source_indices[i]
    dist_matrix[j_source] = 0
    heap.push(dist_index_pair_t(-dist_matrix[j_source], j_source))
    if return_sources:
        sources[j_source] = j_source
```

and add the following code instead,

```python
dist_matrix[source_index] = 0
heap.push(dist_index_pair_t(0, source_index))
if return_sources:
    sources[source_index] = source_index
```

However this change gave reduction in performance (very suprising for me but anyways).

**So I think overall, this PR is at its best and should be merged as is. There can't be more performance gains (there is no recursion in the algorithm which can be converted into iterative loops). Therefore I am approving at my end.** Excellent work @tsery-ns